reader.go

```diff
@@ -151,65 +179,67 @@ func (rd *Reader) readHeader() (*Header, error) {
 		rd.pad = 0
 	}
 
-	// Handle long filenames
-	if header.Name[0] == '/' && rd.longFilenames != nil {
-		if offset, err := strconv.Atoi(header.Name[1:]); err == nil {
-			data := rd.longFilenames[offset:]
-			if idx := bytes.IndexByte(data, '\n'); idx != -1 {
-				data = data[:idx]
-			}
-			if idx := bytes.IndexByte(data, '/'); idx != -1 {
-				data = data[:idx]
-			}
-			header.Name = string(data)
+	// The ar format only supports file names of up to 16 bytes to be stored in the file name header
+	// field. Various archivers have adopted different conventions for storing files with longer
+	// names:
+	// - GNU ar replaces the file name in the header with the character "/" followed by an integer. The
+	//   integer is a byte offset into the archive's symbol table - by convention the archive's first
+	//   file (or second, if a symbol table is also present) - that contains the archive's long file
+	//   names delimited by newlines.
+	if header.Name == "//" {
+		if rd.stringTable != nil {
+			return nil, ErrManyStringTables
+		}
+		buf := make([]byte, rd.nb)
+		_, err := rd.Read(buf)
+		if err != nil {
+			return nil, fmt.Errorf("ar: read GNU long file name data: %w", err)
+		}
+		rd.stringTable = buf
+		// The string table should be invisible to the caller - return the header for the first real file
+		// in the archive.
+		return rd.Next()
+	} else if header.Name[0] == '/' {
+		start, err := strconv.Atoi(header.Name[1:])
+		if err != nil {
+			return nil, fmt.Errorf("ar: invalid GNU long file name offset '%s'", header.Name[1:])
 		}
+		fileName := rd.stringTable[start:]
+		end := bytes.IndexByte(fileName, '\n')
```

Comment:
Do we need to handle the possibility of `\n` being a character in the actual filename?